File: 7/py/tree.py
class DataTree:
    def __init__(self, name, parent=None):
        self.name = name
        self.parent = parent
        self.children = {}
        self.files = {}

    def mkdir(self, name):
        self.children[name] = DataTree(name, parent=self)
        return self.children[name]

    def touch(self, size, name):
        self.files[name] = int(size)

    # stolen hehe c:>

    @property
    def root(self):
        return self if self.parent is None else self.parent.root

    @property
    def size(self):
        return sum(self.files.values()) + sum(child.size for child in self.children.values())

    def __iter__(self):
        for child in self.children.values():
            yield child
            yield from child

File: 7/rb/tree.rb
class Tree:
    @@name
    @@parent
    @@children = {}
    @@files = {}
    
    def initialize(name, parent=nil)
        @@name = name
        @@parent = parent
    end

    def mkdir(name)
        @@children[name] = new Tree(name, self)
        return @@children[name]
    end

    def touch(size, name)
        files[name] = size.to_i
    end

    def root
        nil
    end

    def size
        nil
    end
end